Kubernetes/legos/k8s_get_memory_utilization_of_services/k8s_get_memory_utilization_of_services.py

```diff
@@ -118,39 +82,62 @@ def k8s_get_memory_utilization_of_services(handle, namespace: str = "", threshol
 
     if services and not namespace:
         raise ValueError("Namespace must be provided if services are specified.")
-    
+
     if not namespace:
         namespace = 'default'
 
     exceeding_services = []
+
+    # Main Idea:
+    # 1. Given namespace, lets get current memory utilization for top pods
+    # 2. Filter the list of pods to check from the service list
+    # 3. For the pods get the memory request
+    # 4. Calculate utilization as (mem_usage / mem_request) * 100
+    # 5. Construct list of pods which has  Utilization > threshold  and return the list
+
     try:
-        json_data = {}
-        if not services:
-            kubectl_command = f"kubectl get pods -n {namespace} -o json"
+        # Get memory utilization of all pods using top pods
+        top_pods_command = f"kubectl top pods -n {namespace} --no-headers"
+        response = handle.run_native_cmd(top_pods_command)
+        top_pods_output = response.stdout.strip() 
+        if not top_pods_output:
+            raise ValueError(f"Top PODS data is empty for given namespace {namespace}")
+        top_pods_output = top_pods_output.split('\n')
+        pod_mem_util_dict = {x.split()[0]: x.split()[-1] for x in top_pods_output}
+
+        pods_to_check = {}
+        if services:
+            # If services pod specified, lets iterate over it and check only for them.
+            # If the mentioned pod not found in the top pod list, which means the memory
+            # utilization is not significant, so dont need to check
+            for svc_pod in services:
+                if svc_pod in pod_mem_util_dict.keys():
+                    pods_to_check[svc_pod] = pod_mem_util_dict[svc_pod]
+
+        if not pods_to_check:
+            pods_to_check = pod_mem_util_dict
+
+        for pod, mem_usage in pods_to_check.items():
+            kubectl_command = f"kubectl get pod {pod} -n {namespace} -o=jsonpath='{{.spec.containers[0].resources.requests.memory}}'"
```

Comment:
What if this field is not specified for a pod? what happens in that case?

Reply:
If the field is not specified the mem_request will return as None or empty, so the check `if not mem_request` will become true and the loop continues to the next pod to check.